homeassistant/components/weather/environment_canada.py

```diff
@@ -0,0 +1,228 @@
+"""
+Platform for retrieving meteorological data from Environment Canada.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/weather.environmentcanada/
+"""
+import datetime
+import logging
+import re
+
+from env_canada import ECData
+import voluptuous as vol
+
+from homeassistant.components.weather import (
+    ATTR_FORECAST_CONDITION, ATTR_FORECAST_TEMP, ATTR_FORECAST_TEMP_LOW,
+    ATTR_FORECAST_TIME, PLATFORM_SCHEMA, WeatherEntity)
+from homeassistant.const import (
+    CONF_LATITUDE, CONF_LONGITUDE, CONF_NAME, TEMP_CELSIUS)
+from homeassistant.util import Throttle
+import homeassistant.util.dt as dt
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['env_canada==0.0.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_FORECAST = 'forecast'
+CONF_ATTRIBUTION = "Data provided by Environment Canada"
+CONF_STATION = 'station'
+
+MIN_TIME_BETWEEN_UPDATES = datetime.timedelta(minutes=10)
+
+
+def validate_station(station):
+    """Check that the station ID is well-formed."""
+    if station is None:
+        return
+    if not re.fullmatch(r'[A-Z]{2}/s0000\d{3}', station):
+        raise vol.error.Invalid('Station ID must be of the form "XX/s0000###"')
+    return station
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_STATION): validate_station,
+    vol.Inclusive(CONF_LATITUDE, 'latlon'): cv.latitude,
+    vol.Inclusive(CONF_LONGITUDE, 'latlon'): cv.longitude,
+    vol.Required(CONF_FORECAST, default='daily'):
+        vol.In(['daily', 'hourly']),
+})
+
+# Icon codes from http://dd.weatheroffice.ec.gc.ca/citypage_weather/
+# docs/current_conditions_icon_code_descriptions_e.csv
+ICON_CONDITION_MAP = {'sunny': [0, 1],
+                      'clear-night': [30, 31],
+                      'partlycloudy': [2, 3, 4, 5, 22, 32, 33, 34, 35],
+                      'cloudy': [10],
+                      'rainy': [6, 9, 11, 12, 28, 36],
+                      'lightning-rainy': [19, 39, 46, 47],
+                      'pouring': [13],
+                      'snowy-rainy': [7, 14, 15, 27, 37],
+                      'snowy': [8, 16, 17, 18, 25, 26, 38, 40],
+                      'windy': [43],
+                      'fog': [20, 21, 23, 24, 44],
+                      'hail': [26, 27]}
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Environment Canada weather."""
+    if config.get(CONF_STATION):
```

Comment:
`CONF_STATION`, `CONF_LONGITUDE`, `CONF_LATITUDE` should move to your component, 
`ec_data` should be stored in `hass.data['environment_canada']`
Then in setup_platform, you only need `add_devices([ECWeather(hass.data['environment_canada'], config)])`

Reply:
Okay, I think I see what you mean. In that case, would it still be possible to instantiate the individual platforms without the component? I expect that some users may only want the weather platform, and not the sensor or camera.

Reply:
The component is always required. It's the platforms that can be optional. There are two ways:

1) Have all the config on the component section and load platforms via discovery passing platform config options via discovery_info. See:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/rainmachine/__init__.py
2) Have the common config in the component and require separate platform config sections that sets up the platforms. The platform modules will depend on the component. This was the old way of doing things. There are many examples.

Reply:
I've closed the component PR (#15777) and am going to just have separate platforms.